MigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated_withSafetyCheckAfterTerminate

```
java.lang.AssertionError: Replica: 1 is not found in Partition [12]{
    0:Address[127.0.0.1]:10058
    2:Address[127.0.0.1]:10056
    3:Address[127.0.0.1]:10053
    5:Address[127.0.0.1]:10007
    6:Address[127.0.0.1]:10055
}
    at org.junit.Assert.fail(Assert.java:88)
    at org.junit.Assert.assertTrue(Assert.java:41)
    at org.junit.Assert.assertNotNull(Assert.java:712)
    at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndData(PartitionCorrectnessTestSupport.java:193)
    at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport$AssertSizeAndDataTask.run(PartitionCorrectnessTestSupport.java:247)
    at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:857)
    at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:871)
    at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndDataEventually(PartitionCorrectnessTestSupport.java:150)
    at com.hazelcast.internal.partition.MigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated(MigrationCorrectnessTest.java:122)
    at com.hazelcast.internal.partition.MigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated_withSafetyCheckAfterTerminate(MigrationCorrectnessTest.java:100)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x/com.hazelcast$hazelcast/4966/testReport/junit/com.hazelcast.internal.partition/MigrationCorrectnessTest/testPartitionData_whenBackupNodesStartedTerminated_withSafetyCheckAfterTerminate_backups_3_nodes_7_/

```
java.lang.AssertionError: Replica: 1 is not found in Partition [28]{
	0:[127.0.0.1]:9203
	4:[127.0.0.1]:9206
}
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertNotNull(Assert.java:712)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertPartitionVersionsAndBackupValues(PartitionCorrectnessTestSupport.java:216)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndData(PartitionCorrectnessTestSupport.java:177)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport$AssertSizeAndDataTask.run(PartitionCorrectnessTestSupport.java:291)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:905)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:919)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndDataEventually(PartitionCorrectnessTestSupport.java:152)
	at com.hazelcast.internal.partition.AbstractMigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated(AbstractMigrationCorrectnessTest.java:99)
	at com.hazelcast.internal.partition.AbstractMigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated_withSafetyCheckAfterTerminate(AbstractMigrationCorrectnessTest.java:77)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:88)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:82)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:745)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-ZuluJDK8/com.hazelcast$hazelcast/332/testReport/junit/com.hazelcast.internal.partition/MigrationCorrectnessTest/testPartitionData_whenBackupNodesStartedTerminated_withSafetyCheckAfterTerminate_backups_3_nodes_4_/

```
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.terminateNodes(PartitionCorrectnessTestSupport.java:146)
	at com.hazelcast.internal.partition.AbstractMigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated(AbstractMigrationCorrectnessTest.java:95)
	at com.hazelcast.internal.partition.AbstractMigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated_withSafetyCheckAfterTerminate(AbstractMigrationCorrectnessTest.java:77)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:95)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:56)
	at java.lang.reflect.Method.invoke(Method.java:620)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:88)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:82)
	at java.util.concurrent.FutureTask.run(FutureTask.java:274)
	at java.lang.Thread.run(Thread.java:809)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-IbmJDK1.7/com.hazelcast$hazelcast/1105/testReport/junit/com.hazelcast.internal.partition/MigrationCorrectnessTest/testPartitionData_whenBackupNodesStartedTerminated_withSafetyCheckAfterTerminate_backups_3_nodes_4_/

```
java.lang.AssertionError: Replica: 2 is not found in Partition [2]{
	0:[127.0.0.1]:5072
	1:[127.0.0.1]:5074
}
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertNotNull(Assert.java:712)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertPartitionVersionsAndBackupValues(PartitionCorrectnessTestSupport.java:238)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndData(PartitionCorrectnessTestSupport.java:199)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.access$000(PartitionCorrectnessTestSupport.java:58)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport$AssertSizeAndDataTask.run(PartitionCorrectnessTestSupport.java:326)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:913)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:927)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndDataEventually(PartitionCorrectnessTestSupport.java:170)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndDataEventually(PartitionCorrectnessTestSupport.java:166)
	at com.hazelcast.internal.partition.AbstractMigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated(AbstractMigrationCorrectnessTest.java:93)
	at com.hazelcast.internal.partition.AbstractMigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated_withSafetyCheckAfterTerminate(AbstractMigrationCorrectnessTest.java:77)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
	at java.lang.reflect.Method.invoke(Method.java:597)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:88)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:82)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
	at java.util.concurrent.FutureTask.run(FutureTask.java:138)
	at java.lang.Thread.run(Thread.java:662)
```

https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/6308/testReport/com.hazelcast.internal.partition/MigrationCorrectnessTest/testPartitionData_whenBackupNodesStartedTerminated_withSafetyCheckAfterTerminate_backups_2_nodes_3_/

New failure from the PR builder:

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-pr-builder/16722/testReport/junit/com.hazelcast.internal.partition/MigrationCorrectnessTest/testPartitionData_whenBackupNodesStartedTerminated_withSafetyCheckAfterTerminate_backups_3_nodes_4_fragmented_false_/

**Error Message**
```
Replica version mismatch! Owner: [127.0.0.1]:5701, Backup: [127.0.0.1]:5708, Partition: Partition [0]{
 0:[127.0.0.1]:5701
 1:[127.0.0.1]:5708
 2:[127.0.0.1]:5709
}, Replica: 1 owner versions: [1, 1, 1, 0, 0, 0] backup versions: [0, 0, 0, 0, 0, 0] expected:<1> but was:<0>
```
**Stacktrace**
```
java.lang.AssertionError: 
Replica version mismatch! Owner: [127.0.0.1]:5701, Backup: [127.0.0.1]:5708, Partition: Partition [0]{
	0:[127.0.0.1]:5701
	1:[127.0.0.1]:5708
	2:[127.0.0.1]:5709
}, Replica: 1 owner versions: [1, 1, 1, 0, 0, 0] backup versions: [0, 0, 0, 0, 0, 0] expected:<1> but was:<0>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertPartitionVersionsAndBackupValues(PartitionCorrectnessTestSupport.java:284)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndData(PartitionCorrectnessTestSupport.java:220)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.access$000(PartitionCorrectnessTestSupport.java:61)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport$AssertSizeAndDataTask.run(PartitionCorrectnessTestSupport.java:359)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1279)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1305)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndDataEventually(PartitionCorrectnessTestSupport.java:178)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndDataEventually(PartitionCorrectnessTestSupport.java:174)
	at com.hazelcast.internal.partition.AbstractMigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated(AbstractMigrationCorrectnessTest.java:104)
	at com.hazelcast.internal.partition.AbstractMigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated_withSafetyCheckAfterTerminate(AbstractMigrationCorrectnessTest.java:82)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:98)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```
**Some notable accumulated pauses**
```
13:49:45, accumulated pauses: 1353 ms, max pause: 186 ms, pauses over 1000 ms: 0
13:49:50, accumulated pauses: 1031 ms, max pause: 213 ms, pauses over 1000 ms: 0
13:49:55, accumulated pauses: 1611 ms, max pause: 518 ms, pauses over 1000 ms: 0
13:50:00, accumulated pauses: 1231 ms, max pause: 405 ms, pauses over 1000 ms: 0
13:50:05, accumulated pauses: 1092 ms, max pause: 263 ms, pauses over 1000 ms: 0
13:50:10, accumulated pauses: 1054 ms, max pause: 179 ms, pauses over 1000 ms: 0
13:50:15, accumulated pauses: 800 ms, max pause: 248 ms, pauses over 1000 ms: 0
13:50:20, accumulated pauses: 368 ms, max pause: 9 ms, pauses over 1000 ms: 0
13:50:25, accumulated pauses: 394 ms, max pause: 13 ms, pauses over 1000 ms: 0
13:50:30, accumulated pauses: 491 ms, max pause: 178 ms, pauses over 1000 ms: 0
13:50:35, accumulated pauses: 306 ms, max pause: 13 ms, pauses over 1000 ms: 0
13:50:40, accumulated pauses: 315 ms, max pause: 10 ms, pauses over 1000 ms: 0
13:50:45, accumulated pauses: 326 ms, max pause: 27 ms, pauses over 1000 ms: 0
13:50:50, accumulated pauses: 328 ms, max pause: 16 ms, pauses over 1000 ms: 0
13:50:55, accumulated pauses: 249 ms, max pause: 18 ms, pauses over 1000 ms: 0
13:51:00, accumulated pauses: 367 ms, max pause: 55 ms, pauses over 1000 ms: 0
13:51:05, accumulated pauses: 295 ms, max pause: 25 ms, pauses over 1000 ms: 0
13:51:10, accumulated pauses: 219 ms, max pause: 7 ms, pauses over 1000 ms: 0
13:51:15, accumulated pauses: 389 ms, max pause: 25 ms, pauses over 1000 ms: 0
13:51:20, accumulated pauses: 118 ms, max pause: 6 ms, pauses over 1000 ms: 0
13:51:25, accumulated pauses: 108 ms, max pause: 7 ms, pauses over 1000 ms: 0
13:51:30, accumulated pauses: 89 ms, max pause: 3 ms, pauses over 1000 ms: 0
13:51:35, accumulated pauses: 82 ms, max pause: 9 ms, pauses over 1000 ms: 0
13:51:40, accumulated pauses: 75 ms, max pause: 6 ms, pauses over 1000 ms: 0
13:51:45, accumulated pauses: 64 ms, max pause: 5 ms, pauses over 1000 ms: 0
13:51:50, accumulated pauses: 84 ms, max pause: 8 ms, pauses over 1000 ms: 0
13:51:55, accumulated pauses: 63 ms, max pause: 4 ms, pauses over 1000 ms: 0
13:52:00, accumulated pauses: 51 ms, max pause: 5 ms, pauses over 1000 ms: 0
13:52:05, accumulated pauses: 77 ms, max pause: 3 ms, pauses over 1000 ms: 0
13:52:10, accumulated pauses: 73 ms, max pause: 9 ms, pauses over 1000 ms: 0
13:52:15, accumulated pauses: 55 ms, max pause: 5 ms, pauses over 1000 ms: 0
13:52:20, accumulated pauses: 48 ms, max pause: 4 ms, pauses over 1000 ms: 0
13:52:25, accumulated pauses: 63 ms, max pause: 5 ms, pauses over 1000 ms: 0
13:52:30, accumulated pauses: 45 ms, max pause: 2 ms, pauses over 1000 ms: 0
13:52:35, accumulated pauses: 44 ms, max pause: 6 ms, pauses over 1000 ms: 0
13:52:40, accumulated pauses: 191 ms, max pause: 166 ms, pauses over 1000 ms: 0
```

It failed again on OracleJDK7: http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-3.x-OracleJDK7/8/testReport/com.hazelcast.internal.partition/MigrationCorrectnessTest/testPartitionData_whenBackupNodesStartedTerminated_withSafetyCheckAfterTerminate_backups_3_nodes_4_fragmented_true_/

Stacktrace:
```
java.lang.AssertionError: 
Replica version mismatch! Owner: [127.0.0.1]:5706, Backup: [127.0.0.1]:5716, Partition: Partition [92]{
	0:[127.0.0.1]:5706 - 408d259f-3a58-4d95-ad92-f32ac67a7559
	1:[127.0.0.1]:5710 - 70d1c259-7804-4e23-88a4-9502267819d4
	2:[127.0.0.1]:5717 - b86e50c3-0f35-46d6-9f88-af0c056b18a8
	3:[127.0.0.1]:5716 - 6bdadd52-74af-4090-a48c-27cdf5f2983e
	4:[127.0.0.1]:5714 - 1ca3ff75-35a9-42f0-a6c4-727d532625c9
	5:[127.0.0.1]:5709 - 6a957991-c1ba-47c1-8e48-e2ffd73e1a22
	6:[127.0.0.1]:5715 - a89434be-80df-451d-be65-684116b50252
}, Replica: 3 owner versions: [1, 1, 1, 0, 0, 0] backup versions: [0, 0, 0, 0, 0, 0] expected:<1> but was:<0>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertPartitionVersionsAndBackupValues(PartitionCorrectnessTestSupport.java:284)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndData(PartitionCorrectnessTestSupport.java:222)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.access$000(PartitionCorrectnessTestSupport.java:61)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport$AssertSizeAndDataTask.run(PartitionCorrectnessTestSupport.java:359)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1302)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1328)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndDataEventually(PartitionCorrectnessTestSupport.java:178)
	at com.hazelcast.internal.partition.PartitionCorrectnessTestSupport.assertSizeAndDataEventually(PartitionCorrectnessTestSupport.java:174)
	at com.hazelcast.internal.partition.AbstractMigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated(AbstractMigrationCorrectnessTest.java:98)
	at com.hazelcast.internal.partition.AbstractMigrationCorrectnessTest.testPartitionData_whenBackupNodesStartedTerminated_withSafetyCheckAfterTerminate(AbstractMigrationCorrectnessTest.java:82)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:98)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.lang.Thread.run(Thread.java:745)
```

Accumulated pauses:
```
Hiccups measured while running test 'testPartitionData_whenBackupNodesStartedTerminated_withSafetyCheckAfterTerminate[backups:3,nodes:4,fragmented:true](com.hazelcast.internal.partition.MigrationCorrectnessTest):'
10:31:15, accumulated pauses: 143 ms, max pause: 33 ms, pauses over 1000 ms: 0
10:31:20, accumulated pauses: 156 ms, max pause: 28 ms, pauses over 1000 ms: 0
10:31:25, accumulated pauses: 61 ms, max pause: 21 ms, pauses over 1000 ms: 0
10:31:30, accumulated pauses: 40 ms, max pause: 0 ms, pauses over 1000 ms: 0
10:31:35, accumulated pauses: 64 ms, max pause: 24 ms, pauses over 1000 ms: 0
10:31:40, accumulated pauses: 39 ms, max pause: 0 ms, pauses over 1000 ms: 0
10:31:45, accumulated pauses: 54 ms, max pause: 15 ms, pauses over 1000 ms: 0
10:31:50, accumulated pauses: 38 ms, max pause: 0 ms, pauses over 1000 ms: 0
10:31:55, accumulated pauses: 40 ms, max pause: 0 ms, pauses over 1000 ms: 0
10:32:00, accumulated pauses: 48 ms, max pause: 7 ms, pauses over 1000 ms: 0
10:32:05, accumulated pauses: 39 ms, max pause: 0 ms, pauses over 1000 ms: 0
10:32:10, accumulated pauses: 53 ms, max pause: 12 ms, pauses over 1000 ms: 0
10:32:15, accumulated pauses: 40 ms, max pause: 0 ms, pauses over 1000 ms: 0
10:32:20, accumulated pauses: 37 ms, max pause: 0 ms, pauses over 1000 ms: 0
10:32:25, accumulated pauses: 53 ms, max pause: 14 ms, pauses over 1000 ms: 0
10:32:30, accumulated pauses: 39 ms, max pause: 0 ms, pauses over 1000 ms: 0
10:32:35, accumulated pauses: 44 ms, max pause: 6 ms, pauses over 1000 ms: 0
10:32:40, accumulated pauses: 41 ms, max pause: 0 ms, pauses over 1000 ms: 0
10:32:45, accumulated pauses: 39 ms, max pause: 0 ms, pauses over 1000 ms: 0
10:32:50, accumulated pauses: 51 ms, max pause: 11 ms, pauses over 1000 ms: 0
10:32:55, accumulated pauses: 41 ms, max pause: 0 ms, pauses over 1000 ms: 0
10:33:00, accumulated pauses: 57 ms, max pause: 18 ms, pauses over 1000 ms: 0
10:33:05, accumulated pauses: 40 ms, max pause: 0 ms, pauses over 1000 ms: 0
10:33:10, accumulated pauses: 38 ms, max pause: 0 ms, pauses over 1000 ms: 0
10:33:15, accumulated pauses: 53 ms, max pause: 11 ms, pauses over 1000 ms: 0
10:33:20, accumulated pauses: 42 ms, max pause: 0 ms, pauses over 1000 ms: 0
10:33:25, accumulated pauses: 53 ms, max pause: 12 ms, pauses over 1000 ms: 0
10:33:30, accumulated pauses: 39 ms, max pause: 0 ms, pauses over 1000 ms: 0
10:33:35, accumulated pauses: 39 ms, max pause: 0 ms, pauses over 1000 ms: 0
10:33:40, accumulated pauses: 56 ms, max pause: 15 ms, pauses over 1000 ms: 0
10:33:45, accumulated pauses: 38 ms, max pause: 0 ms, pauses over 1000 ms: 0
10:33:50, accumulated pauses: 754 ms, max pause: 721 ms, pauses over 1000 ms: 0
```

Debug log was enabled for that run. I also marked that run as `Keep this build forever` to persist those logs in Jenkins. Once when investigation will be done then please set that run back to `Don't keep this build forever`.

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/MigrationInfo.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/MigrationInfo.java
@@ -79,6 +79,8 @@ public class MigrationInfo implements IdentifiedDataSerializable, Versioned {
     private int sourceNewReplicaIndex;
     private int destinationCurrentReplicaIndex;
     private int destinationNewReplicaIndex;
+    private int initialPartitionVersion = -1;
+    private int partitionVersionIncrement;
 
     private final AtomicBoolean processing = new AtomicBoolean(false);
     private volatile MigrationStatus status;
@@ -108,20 +110,12 @@ public class MigrationInfo implements IdentifiedDataSerializable, Versioned {
         return source != null ? source.address() : null;
     }
 
-    public String getSourceUuid() {
-        return source != null ? source.uuid() : null;
-    }
-
     public PartitionReplica getDestination() {
         return destination;
     }
 
     public Address getDestinationAddress() {
         return destination != null ? destination.address() : null;
-    }
-
-    public String getDestinationUuid() {
-        return destination != null ? destination.uuid() : null;
     }
 
     public int getPartitionId() {
@@ -157,10 +151,6 @@ public class MigrationInfo implements IdentifiedDataSerializable, Versioned {
         return processing.compareAndSet(false, true);
     }
 
-    public boolean isProcessing() {
-        return processing.get();
-    }
-
     public void doneProcessing() {
         processing.set(false);
     }
@@ -178,6 +168,43 @@ public class MigrationInfo implements IdentifiedDataSerializable, Versioned {
         return status != MigrationStatus.INVALID;
     }
 
+    public int getInitialPartitionVersion() {
+        return initialPartitionVersion;
+    }
+
+    public MigrationInfo setInitialPartitionVersion(int initialPartitionVersion) {
+        assert initialPartitionVersion > 0;
+        this.initialPartitionVersion = initialPartitionVersion;
+        return this;
+    }
+
+    public int getPartitionVersionIncrement() {
+        if (partitionVersionIncrement > 0) {
+            return partitionVersionIncrement;
+        }
+        int inc = 1;
+        if (sourceNewReplicaIndex > -1) {
+            inc++;
+        }
+        if (destinationCurrentReplicaIndex > -1) {
+            inc++;
+        }
+        return inc;
+    }
+
+    public MigrationInfo setPartitionVersionIncrement(int partitionVersionIncrement) {
+        assert partitionVersionIncrement > 0;
+        this.partitionVersionIncrement = partitionVersionIncrement;
+        return this;
+    }
+
+    public int getFinalPartitionVersion() {
+        if (initialPartitionVersion > 0) {
+            return initialPartitionVersion + getPartitionVersionIncrement();
+        }
+        throw new IllegalStateException("Initial partition version is not set!");
+    }
+
     @Override
     public void writeData(ObjectDataOutput out) throws IOException {
         out.writeUTF(uuid);
@@ -193,6 +220,7 @@ public class MigrationInfo implements IdentifiedDataSerializable, Versioned {
         boolean hasSource = source != null;
         out.writeBoolean(hasSource);
         if (hasSource) {
+            // RU_COMPAT_3_11
             if (version.isGreaterOrEqual(Versions.V3_12)) {
                 out.writeObject(source);
             } else {
@@ -203,6 +231,7 @@ public class MigrationInfo implements IdentifiedDataSerializable, Versioned {
         boolean hasDestination = destination != null;
         out.writeBoolean(hasDestination);
         if (hasDestination) {
+            // RU_COMPAT_3_11
             if (version.isGreaterOrEqual(Versions.V3_12)) {
                 out.writeObject(destination);
             } else {
@@ -211,6 +240,12 @@ public class MigrationInfo implements IdentifiedDataSerializable, Versioned {
         }
 
         master.writeData(out);
+
+        // RU_COMPAT_3_11
+        if (version.isGreaterOrEqual(Versions.V3_12)) {
+            out.writeInt(initialPartitionVersion);
+            out.writeInt(partitionVersionIncrement);
+        }
     }
 
     private static void writePartitionReplicaLegacy(ObjectDataOutput out, PartitionReplica destination) throws IOException {
@@ -231,6 +266,7 @@ public class MigrationInfo implements IdentifiedDataSerializable, Versioned {
         Version version = in.getVersion();
         boolean hasSource = in.readBoolean();
         if (hasSource) {
+            // RU_COMPAT_3_11
             if (version.isGreaterOrEqual(Versions.V3_12)) {
                 source = in.readObject();
             } else {
@@ -240,6 +276,7 @@ public class MigrationInfo implements IdentifiedDataSerializable, Versioned {
 
         boolean hasDestination = in.readBoolean();
         if (hasDestination) {
+            // RU_COMPAT_3_11
             if (version.isGreaterOrEqual(Versions.V3_12)) {
                 destination = in.readObject();
             } else {
@@ -249,6 +286,12 @@ public class MigrationInfo implements IdentifiedDataSerializable, Versioned {
 
         master = new Address();
         master.readData(in);
+
+        // RU_COMPAT_3_11
+        if (version.isGreaterOrEqual(Versions.V3_12)) {
+            initialPartitionVersion = in.readInt();
+            partitionVersionIncrement = in.readInt();
+        }
     }
 
     private static PartitionReplica readPartitionReplicaLegacy(ObjectDataInput in) throws IOException {
@@ -290,6 +333,8 @@ public class MigrationInfo implements IdentifiedDataSerializable, Versioned {
         sb.append(", destinationCurrentReplicaIndex=").append(destinationCurrentReplicaIndex);
         sb.append(", destinationNewReplicaIndex=").append(destinationNewReplicaIndex);
         sb.append(", master=").append(master);
+        sb.append(", initialPartitionVersion=").append(initialPartitionVersion);
+        sb.append(", partitionVersionIncrement=").append(getPartitionVersionIncrement());
         sb.append(", processing=").append(processing);
         sb.append(", status=").append(status);
         sb.append('}');

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/InternalPartitionServiceImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/InternalPartitionServiceImpl.java
@@ -74,7 +74,6 @@ import com.hazelcast.util.scheduler.CoalescingDelayedTrigger;
 import com.hazelcast.util.scheduler.ScheduledEntry;
 import com.hazelcast.version.Version;
 
-import javax.annotation.Nullable;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
@@ -534,7 +533,8 @@ public class InternalPartitionServiceImpl implements InternalPartitionService,
                 migrationInfo.setStatus(MigrationStatus.SUCCESS);
             }
 
-            int committedVersion = getPartitionStateVersion() + migrationInfos.size();
+            // each promotion increments version by 2
+            int committedVersion = getPartitionStateVersion() + migrationInfos.size() * 2;
             return new PartitionRuntimeState(partitions, completedMigrations, committedVersion);
         } finally {
             lock.unlock();
@@ -585,11 +585,10 @@ public class InternalPartitionServiceImpl implements InternalPartitionService,
     }
 
     void sendPartitionRuntimeState(Address target) {
-        if (!node.isRunning()) {
+        if (!node.isMaster()) {
             return;
         }
         assert partitionStateManager.isInitialized();
-        assert node.isMaster();
         assert areMigrationTasksAllowed();
 
         PartitionRuntimeState partitionState = createPartitionStateInternal();
@@ -766,9 +765,18 @@ public class InternalPartitionServiceImpl implements InternalPartitionService,
         }
 
         try {
-            Boolean versionCheckResult = checkIfPartitionStateVersionApplied(newVersion, sender);
-            if (versionCheckResult != null) {
-                return versionCheckResult;
+            int currentVersion = partitionStateManager.getVersion();
+            if (newVersion < currentVersion) {
+                if (logger.isFineEnabled()) {
+                    logger.fine("Already applied partition state change. Local version: " + currentVersion
+                            + ", Master version: " + newVersion + " Master: " + sender);
+                }
+                return false;
+            } else if (newVersion == currentVersion) {
+                if (logger.isFineEnabled()) {
+                    logger.fine("Already applied partition state change. Version: " + currentVersion + ", Master: " + sender);
+                }
+                return true;
             }
 
             // RU_COMPAT_3_11
@@ -814,31 +822,6 @@ public class InternalPartitionServiceImpl implements InternalPartitionService,
                 }
             }
         }
-    }
-
-    /**
-     * Checks whether {@code version} is already applied or a new version.
-     *
-     * @return {@code true} if current version is equal to the {@code version},
-     *          {@code false} if {@code version} is stale,
-     *          or {@code null} if {@code version} is newer and should be applied.
-     */
-    @Nullable
-    private Boolean checkIfPartitionStateVersionApplied(int version, Address sender) {
-        int currentVersion = partitionStateManager.getVersion();
-        if (version < currentVersion) {
-            if (logger.isFineEnabled()) {
-                logger.fine("Already applied partition state change. Local version: " + currentVersion
-                        + ", Master version: " + version + " Master: " + sender);
-            }
-            return false;
-        } else if (version == currentVersion) {
-            if (logger.isFineEnabled()) {
-                logger.fine("Already applied partition state change. Version: " + currentVersion + ", Master: " + sender);
-            }
-            return true;
-        }
-        return null;
     }
 
     private void requestMemberListUpdateIfUnknownMembersFound(Address sender, PartitionReplica[][] partitionTable) {
@@ -898,59 +881,26 @@ public class InternalPartitionServiceImpl implements InternalPartitionService,
             partitionStateManager.updateReplicas(partitionId, replicas);
         }
 
-        applyCompletedMigrationsInternal(completedMigrations, version);
-    }
+        partitionStateManager.setVersion(version);
 
-    /**
-     * Applies completed migrations to the partition table, sets partition state version
-     * and schedules finalization step of completed migrations.
-     *
-     * @param completedMigrations completed migrations
-     * @param newVersion new partition state version
-     */
-    private void applyCompletedMigrationsInternal(Collection<MigrationInfo> completedMigrations, int newVersion) {
-        lock.lock();
-        try {
-            Collection<MigrationInfo> recentlyCompletedMigrations = new LinkedList<MigrationInfo>();
-            for (MigrationInfo migration : completedMigrations) {
-                assert migration.getStatus() == MigrationStatus.SUCCESS
-                        || migration.getStatus() == MigrationStatus.FAILED
-                        : "Invalid migration: " + migration;
-
-                boolean added = migrationManager.addCompletedMigration(migration);
-                if (added) {
-                    recentlyCompletedMigrations.add(migration);
-
-                    if (migration.getStatus() == MigrationStatus.SUCCESS) {
-                        if (logger.isFinestEnabled()) {
-                            logger.finest("Applying completed migration " + migration);
-                        }
-                        InternalPartitionImpl partition = partitionStateManager.getPartitionImpl(migration.getPartitionId());
-                        migrationManager.applyMigration(partition, migration);
-                    }
-                }
-            }
-            partitionStateManager.setVersion(newVersion);
-            if (logger.isFinestEnabled()) {
-                logger.finest("Applied completed migrations with partition state version: " + newVersion);
-            }
-
-            if (!partitionStateManager.setInitialized()) {
-                node.getNodeExtension().onPartitionStateChange();
-            }
-
-            migrationManager.retainCompletedMigrations(completedMigrations);
-
-            for (MigrationInfo migration : recentlyCompletedMigrations) {
+        for (MigrationInfo migration : completedMigrations) {
+            boolean added = migrationManager.addCompletedMigration(migration);
+            if (added) {
                 migrationManager.scheduleActiveMigrationFinalization(migration);
             }
+        }
+        if (logger.isFineEnabled()) {
+            logger.fine("Applied partition state update with version: " + version);
+        }
+        migrationManager.retainCompletedMigrations(completedMigrations);
 
-        } finally {
-            lock.unlock();
+        if (!partitionStateManager.setInitialized()) {
+            node.getNodeExtension().onPartitionStateChange();
         }
     }
 
-    public boolean applyCompletedMigrations(Collection<MigrationInfo> migrations, int newVersion, Address sender) {
+    @SuppressWarnings({"checkstyle:cyclomaticcomplexity", "checkstyle:npathcomplexity"})
+    public boolean applyCompletedMigrations(Collection<MigrationInfo> migrations, Address sender) {
         if (!validateSenderIsMaster(sender, "completed migrations")) {
             return false;
         }
@@ -959,17 +909,53 @@ public class InternalPartitionServiceImpl implements InternalPartitionService,
             if (!partitionStateManager.isInitialized()) {
                 if (logger.isFineEnabled()) {
                     logger.fine("Cannot apply completed migrations until partition table is initialized. "
-                            + "Partition state version: " + newVersion);
+                            + "Completed migrations: " + migrations);
                 }
                 return false;
             }
-            Boolean versionCheckResult = checkIfPartitionStateVersionApplied(newVersion, sender);
-            if (versionCheckResult != null) {
-                return versionCheckResult;
+
+            boolean appliedAllMigrations = true;
+            for (MigrationInfo migration : migrations) {
+                int currentVersion = partitionStateManager.getVersion();
+                if (migration.getFinalPartitionVersion() <= currentVersion) {
+                    if (logger.isFinestEnabled()) {
+                        logger.finest("Already applied migration commit. Local version: " + currentVersion
+                                + ", Commit version: " + migration.getFinalPartitionVersion() + " Master: " + sender);
+                    }
+                    continue;
+                }
+
+                if (migration.getInitialPartitionVersion() != currentVersion) {
+                    logger.fine("Cannot apply migration commit! Expected version: " + migration.getInitialPartitionVersion()
+                            + ", current version: " + currentVersion + ", final version: " + migration.getFinalPartitionVersion()
+                            + ", Master: " + sender);
+                    appliedAllMigrations = false;
+                    break;
+                }
+
+                boolean added = migrationManager.addCompletedMigration(migration);
+                assert added : "Migration: " + migration;
+
+                partitionStateManager.incrementVersion(migration.getPartitionVersionIncrement());
+
+                if (migration.getStatus() == MigrationStatus.SUCCESS) {
+                    if (logger.isFineEnabled()) {
+                        logger.fine("Applying completed migration " + migration);
+                    }
+                    InternalPartitionImpl partition = partitionStateManager.getPartitionImpl(migration.getPartitionId());
+                    migrationManager.applyMigration(partition, migration);
+                    migrationManager.scheduleActiveMigrationFinalization(migration);
+                }
             }
 
-            applyCompletedMigrationsInternal(migrations, newVersion);
-            return true;
+            if (logger.isFineEnabled()) {
+                logger.fine("Applied completed migrations with partition state version: " + partitionStateManager.getVersion());
+            }
+
+            migrationManager.retainCompletedMigrations(migrations);
+            node.getNodeExtension().onPartitionStateChange();
+
+            return appliedAllMigrations;
         } finally {
             lock.unlock();
         }
@@ -1334,36 +1320,54 @@ public class InternalPartitionServiceImpl implements InternalPartitionService,
         }
     }
 
-    public boolean commitMigrationOnDestination(MigrationInfo migration, int newVersion, Address sender) {
+    @SuppressWarnings("checkstyle:npathcomplexity")
+    public boolean commitMigrationOnDestination(MigrationInfo migration, Address sender) {
         lock.lock();
         try {
             if (!validateSenderIsMaster(sender, "migration commit")) {
                 return false;
             }
 
-            Boolean versionCheckResult = checkIfPartitionStateVersionApplied(newVersion, sender);
-            if (versionCheckResult != null) {
-                return versionCheckResult;
-            }
+            int currentVersion = partitionStateManager.getVersion();
+            int initialVersion = migration.getInitialPartitionVersion();
+            int finalVersion = migration.getFinalPartitionVersion();
 
-            MigrationInfo activeMigrationInfo = migrationManager.getActiveMigration();
-            if (migration.equals(activeMigrationInfo)) {
-                InternalPartitionImpl partition = partitionStateManager.getPartitionImpl(migration.getPartitionId());
-                boolean added = migrationManager.addCompletedMigration(migration);
-                assert added : "Could not add completed migration on destination: " + migration;
-                migrationManager.applyMigration(partition, migration);
-                partitionStateManager.setVersion(newVersion);
-                activeMigrationInfo.setStatus(migration.getStatus());
-                migrationManager.finalizeMigration(migration);
-                if (logger.isFinestEnabled()) {
-                    logger.finest("Committed " + migration + " on destination with partition state version: " + newVersion);
+            if (finalVersion == currentVersion) {
+                if (logger.isFineEnabled()) {
+                    logger.fine("Already applied migration commit. Version: " + currentVersion + ", Master: " + sender);
                 }
                 return true;
             }
+            if (finalVersion < currentVersion) {
+                if (logger.isFineEnabled()) {
+                    logger.fine("Already applied migration commit. Local version: " + currentVersion
+                            + ", Master version: " + finalVersion + " Master: " + sender);
+                }
+                return false;
+            }
+            if (initialVersion != currentVersion) {
+                throw new IllegalStateException("Invalid migration commit! Expected version: " + initialVersion
+                        + ", current version: " + currentVersion + ", Master: " + sender);
+            }
+
+            MigrationInfo activeMigration = migrationManager.getActiveMigration();
+            assert migration.equals(activeMigration) : "Committed migration: " + migration
+                    + ", Active migration: " + activeMigration;
+
+            InternalPartitionImpl partition = partitionStateManager.getPartitionImpl(migration.getPartitionId());
+            boolean added = migrationManager.addCompletedMigration(migration);
+            assert added : "Could not add completed migration on destination: " + migration;
+            migrationManager.applyMigration(partition, migration);
+            partitionStateManager.setVersion(finalVersion);
+            activeMigration.setStatus(migration.getStatus());
+            migrationManager.finalizeMigration(migration);
+            if (logger.isFineEnabled()) {
+                logger.fine("Committed " + migration + " on destination with partition state version: " + finalVersion);
+            }
+            return true;
         } finally {
             lock.unlock();
         }
-        return false;
     }
 
 

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java
@@ -52,6 +52,7 @@ import com.hazelcast.spi.Operation;
 import com.hazelcast.spi.OperationService;
 import com.hazelcast.spi.exception.TargetNotMemberException;
 import com.hazelcast.spi.impl.NodeEngineImpl;
+import com.hazelcast.spi.impl.operationservice.InternalOperationService;
 import com.hazelcast.spi.partition.IPartitionLostEvent;
 import com.hazelcast.spi.partition.MigrationEndpoint;
 import com.hazelcast.spi.properties.GroupProperty;
@@ -64,6 +65,7 @@ import com.hazelcast.version.Version;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Iterator;
@@ -90,7 +92,6 @@ import static com.hazelcast.spi.partition.IPartitionService.SERVICE_NAME;
 @SuppressWarnings({"checkstyle:classdataabstractioncoupling", "checkstyle:methodcount", "checkstyle:classfanoutcomplexity"})
 public class MigrationManager {
 
-    private static final int PARTITION_STATE_VERSION_INCREMENT_DELTA_ON_MIGRATION_FAILURE = 2;
     private static final int MIGRATION_PAUSE_DURATION_SECONDS_ON_MIGRATION_FAILURE = 3;
     private static final int PUBLISH_COMPLETED_MIGRATIONS_BATCH_SIZE = 10;
 
@@ -216,10 +217,19 @@ public class MigrationManager {
 
                 MigrationEndpoint endpoint = source ? MigrationEndpoint.SOURCE : MigrationEndpoint.DESTINATION;
                 FinalizeMigrationOperation op = new FinalizeMigrationOperation(migrationInfo, endpoint, success);
-
                 op.setPartitionId(partitionId).setNodeEngine(nodeEngine).setValidateTarget(false)
                         .setService(partitionService);
-                nodeEngine.getOperationService().execute(op);
+                InternalOperationService operationService = nodeEngine.getOperationService();
+                if (operationService.isRunAllowed(op)) {
+                    // When migration finalization is triggered by subsequent migrations
+                    // on partition thread, finalization may run directly on calling thread.
+                    // Otherwise, if previously completed migration and newly submitted migration
+                    // are on the same partition, then finalizing previous migration later will
+                    // break the order, hence safety.
+                    operationService.run(op);
+                } else {
+                    operationService.execute(op);
+                }
                 removeActiveMigration(partitionId);
             } else {
                 PartitionReplica partitionOwner = partitionStateManager.getPartitionImpl(partitionId).getOwnerReplicaOrNull();
@@ -327,6 +337,10 @@ public class MigrationManager {
                     activeMigrationInfo.setStatus(migrationInfo.getStatus());
                     finalizeMigration(activeMigrationInfo);
                 } else {
+                    // This case happens when master crashes while migration operation is running
+                    // and new master publishes the latest completed migrations
+                    // after collecting active and completed migrations from existing members.
+                    // See FetchMostRecentPartitionTableTask.
                     logger.info("Scheduling finalization of " + migrationInfo
                             + ", because migration process is currently running.");
                     nodeEngine.getExecutionService().schedule(new Runnable() {
@@ -380,9 +394,7 @@ public class MigrationManager {
 
             MigrationCommitOperation operation;
             if (nodeEngine.getClusterService().getClusterVersion().isGreaterOrEqual(Versions.V3_12)) {
-                int newPartitionStateVersion = partitionStateManager.getVersion() + 1;
-                operation = new MigrationCommitOperation(migration, newPartitionStateVersion,
-                        destinationUuid);
+                operation = new MigrationCommitOperation(migration, destinationUuid);
             } else {
                 // RU_COMPAT_3_11
                 PartitionRuntimeState partitionState = partitionService.createMigrationCommitPartitionState(migration);
@@ -440,6 +452,13 @@ public class MigrationManager {
         if (migrationInfo.getStatus() != MigrationStatus.SUCCESS
                 && migrationInfo.getStatus() != MigrationStatus.FAILED) {
             throw new IllegalArgumentException("Migration doesn't seem completed: " + migrationInfo);
+        }
+
+        // RU_COMPAT_3_11
+        if (node.getClusterService().getClusterVersion().isGreaterOrEqual(Versions.V3_12)) {
+            if (migrationInfo.getInitialPartitionVersion() <= 0 || migrationInfo.getPartitionVersionIncrement() <= 0) {
+                throw new IllegalArgumentException("Partition state versions are not set: " + migrationInfo);
+            }
         }
 
         partitionServiceLock.lock();
@@ -647,12 +666,9 @@ public class MigrationManager {
         assert node.isMaster();
         assert partitionStateManager.isInitialized();
 
-        final List<MigrationInfo> migrations = new ArrayList<MigrationInfo>();
-        int partitionStateVersion = fillCompletedMigrationsAndGetPartitionStateVersion(migrations);
-
+        List<MigrationInfo> migrations = getCompletedMigrationsCopy();
         if (logger.isFineEnabled()) {
-            logger.fine("Publishing completed migrations, version: " + partitionStateVersion
-                    + ", migrations: " + migrations);
+            logger.fine("Publishing completed migrations [" + migrations.size() + "]: " + migrations);
         }
 
         OperationService operationService = nodeEngine.getOperationService();
@@ -664,7 +680,7 @@ public class MigrationManager {
             if ((member.localMember())) {
                 continue;
             }
-            Operation operation = new PublishCompletedMigrationsOperation(migrations, partitionStateVersion);
+            Operation operation = new PublishCompletedMigrationsOperation(migrations);
             InternalCompletableFuture<Boolean> f
                     = operationService.invokeOnTarget(SERVICE_NAME, operation, member.getAddress());
 
@@ -689,16 +705,6 @@ public class MigrationManager {
                     partitionService.sendPartitionRuntimeState(member.getAddress());
                 }
             });
-        }
-    }
-
-    private int fillCompletedMigrationsAndGetPartitionStateVersion(Collection<MigrationInfo> migrations) {
-        partitionServiceLock.lock();
-        try {
-            migrations.addAll(completedMigrations);
-            return partitionStateManager.getVersion();
-        } finally {
-            partitionServiceLock.unlock();
         }
     }
 
@@ -821,9 +827,12 @@ public class MigrationManager {
                 migrations.add(migrationCollector.migrations);
             }
             partitionService.publishPartitionRuntimeState();
-            scheduleMigrations(migrations);
-            // Schedule a task to publish completed migrations after all migrations tasks are completed.
-            migrationQueue.add(new PublishCompletedMigrationsTask());
+
+            if (migrationCount.value > 0) {
+                scheduleMigrations(migrations);
+                // Schedule a task to publish completed migrations after all migrations tasks are completed.
+                migrationQueue.add(new PublishCompletedMigrationsTask());
+            }
             logMigrationStatistics(migrationCount.value, lostCount.value);
         }
 
@@ -973,7 +982,7 @@ public class MigrationManager {
                 processMigrationResult(partitionOwner, result);
             } catch (Throwable t) {
                 final Level level = migrationInfo.isValid() ? Level.WARNING : Level.FINE;
-                logger.log(level, "Error [" + t.getClass() + ": " + t.getMessage() + "] during " + migrationInfo);
+                logger.log(level, "Error [" + t.getClass() + ": " + t.getMessage() + "] during " + migrationInfo, t);
                 logger.finest(t);
                 migrationOperationFailed(partitionOwner);
             } finally {
@@ -983,6 +992,7 @@ public class MigrationManager {
 
         /** Sends a migration event to the event listeners. */
         private void beforeMigration() {
+            migrationInfo.setInitialPartitionVersion(partitionStateManager.getVersion());
             internalMigrationListener.onMigrationStart(MigrationParticipant.MASTER, migrationInfo);
             partitionService.getPartitionEventManager()
                     .sendMigrationEvent(migrationInfo, MigrationEvent.MigrationStatus.STARTED);
@@ -1056,8 +1066,12 @@ public class MigrationManager {
          */
         private Boolean executeMigrateOperation(Member fromMember) {
             long start = System.nanoTime();
-            List<MigrationInfo> completedMigrations = new ArrayList<MigrationInfo>();
-            int partitionStateVersion = fillCompletedMigrationsAndGetPartitionStateVersion(completedMigrations);
+            List<MigrationInfo> completedMigrations = Collections.emptyList();
+            // RU_COMPAT_3_11
+            if (node.getClusterService().getClusterVersion().isGreaterOrEqual(Versions.V3_12)) {
+                completedMigrations = getCompletedMigrationsCopy();
+            }
+            int partitionStateVersion = partitionStateManager.getVersion();
             Operation op = new MigrationRequestOperation(migrationInfo, completedMigrations, partitionStateVersion,
                     fragmentedMigrationEnabled);
             Future future = nodeEngine.getOperationService().createInvocationBuilder(SERVICE_NAME, op,
@@ -1103,12 +1117,13 @@ public class MigrationManager {
             internalMigrationListener.onMigrationComplete(MigrationParticipant.MASTER, migrationInfo, false);
             partitionServiceLock.lock();
             try {
-                addCompletedMigration(migrationInfo);
                 internalMigrationListener.onMigrationRollback(MigrationParticipant.MASTER, migrationInfo);
                 scheduleActiveMigrationFinalization(migrationInfo);
-                int delta = PARTITION_STATE_VERSION_INCREMENT_DELTA_ON_MIGRATION_FAILURE;
-                partitionService.getPartitionStateManager().incrementVersion(delta);
+                int delta = migrationInfo.getPartitionVersionIncrement() + 1;
+                partitionStateManager.incrementVersion(delta);
+                migrationInfo.setPartitionVersionIncrement(delta);
                 node.getNodeExtension().onPartitionStateChange();
+                addCompletedMigration(migrationInfo);
 
                 if (!partitionOwner.localMember()) {
                     partitionService.sendPartitionRuntimeState(partitionOwner.getAddress());
@@ -1169,14 +1184,21 @@ public class MigrationManager {
                 if (commitSuccessful) {
                     migrationInfo.setStatus(MigrationStatus.SUCCESS);
                     internalMigrationListener.onMigrationCommit(MigrationParticipant.MASTER, migrationInfo);
+                    assert migrationInfo.getInitialPartitionVersion() == partitionStateManager.getVersion()
+                            : "Migration initial version: " + migrationInfo.getInitialPartitionVersion()
+                            + ", Partition state version: " + partitionStateManager.getVersion();
                     // updates partition table after successful commit
                     InternalPartitionImpl partition = partitionStateManager.getPartitionImpl(migrationInfo.getPartitionId());
                     applyMigration(partition, migrationInfo);
+                    assert migrationInfo.getFinalPartitionVersion() == partitionStateManager.getVersion()
+                            : "Migration final version: " + migrationInfo.getFinalPartitionVersion()
+                                + ", Partition state version: " + partitionStateManager.getVersion();
                 } else {
                     migrationInfo.setStatus(MigrationStatus.FAILED);
                     internalMigrationListener.onMigrationRollback(MigrationParticipant.MASTER, migrationInfo);
-                    int delta = PARTITION_STATE_VERSION_INCREMENT_DELTA_ON_MIGRATION_FAILURE;
-                    partitionService.getPartitionStateManager().incrementVersion(delta);
+                    int delta = migrationInfo.getPartitionVersionIncrement() + 1;
+                    partitionStateManager.incrementVersion(delta);
+                    migrationInfo.setPartitionVersionIncrement(delta);
 
                     if (!migrationInfo.getDestination().isIdentical(node.getLocalMember())) {
                         partitionService.sendPartitionRuntimeState(migrationInfo.getDestination().address());
@@ -1348,11 +1370,12 @@ public class MigrationManager {
                         assert destination.equals(partition.getReplica(migration.getDestinationCurrentReplicaIndex()))
                                 : "Invalid replica! Destination: " + destination + ", index: "
                                         + migration.getDestinationCurrentReplicaIndex() + ", " + partition;
-                        // single partition update increments partition state version by 1
+                        // single partition replica swap, increments partition state version by 2
                         partition.swapReplicas(0, migration.getDestinationCurrentReplicaIndex());
                     }
                 } else {
-                    int delta = migrations.size() + 1;
+                    // each promotion increments version by 2
+                    int delta = 2 * migrations.size() + 1;
                     partitionService.getPartitionStateManager().incrementVersion(delta);
                 }
             } finally {

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionStateManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionStateManager.java
@@ -343,14 +343,11 @@ public class PartitionStateManager {
     }
 
     void incrementVersion(int delta) {
-        if (delta >= 0) {
-            stateVersion.addAndGet(delta);
-        } else {
-            logger.warning("partition table version not incremented by " + delta);
-        }
+        assert delta > 0 : "Delta: " + delta;
+        stateVersion.addAndGet(delta);
     }
 
-    public void incrementVersion() {
+    void incrementVersion() {
         stateVersion.incrementAndGet();
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/BaseMigrationOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/BaseMigrationOperation.java
@@ -112,7 +112,7 @@ abstract class BaseMigrationOperation extends AbstractPartitionOperation
             return;
         }
         InternalPartitionServiceImpl partitionService = getService();
-        if (!partitionService.applyCompletedMigrations(completedMigrations, partitionStateVersion, migrationInfo.getMaster())) {
+        if (!partitionService.applyCompletedMigrations(completedMigrations, migrationInfo.getMaster())) {
             throw new PartitionStateVersionMismatchException(partitionStateVersion, partitionService.getPartitionStateVersion());
         }
     }

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/MigrationCommitOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/MigrationCommitOperation.java
@@ -45,8 +45,6 @@ public class MigrationCommitOperation extends AbstractPartitionOperation impleme
 
     private MigrationInfo migration;
 
-    private int newPartitionStateVersion;
-
     private String expectedMemberUuid;
 
     private transient boolean success;
@@ -60,9 +58,8 @@ public class MigrationCommitOperation extends AbstractPartitionOperation impleme
         this.expectedMemberUuid = expectedMemberUuid;
     }
 
-    public MigrationCommitOperation(MigrationInfo migration, int newPartitionStateVersion, String expectedMemberUuid) {
+    public MigrationCommitOperation(MigrationInfo migration, String expectedMemberUuid) {
         this.migration = migration;
-        this.newPartitionStateVersion = newPartitionStateVersion;
         this.expectedMemberUuid = expectedMemberUuid;
     }
 
@@ -79,7 +76,7 @@ public class MigrationCommitOperation extends AbstractPartitionOperation impleme
         InternalPartitionServiceImpl service = getService();
 
         if (nodeEngine.getClusterService().getClusterVersion().isGreaterOrEqual(Versions.V3_12)) {
-            success = service.commitMigrationOnDestination(migration, newPartitionStateVersion, getCallerAddress());
+            success = service.commitMigrationOnDestination(migration, getCallerAddress());
         } else {
             // RU_COMPAT_3_11
             partitionState.setMaster(getCallerAddress());
@@ -113,7 +110,6 @@ public class MigrationCommitOperation extends AbstractPartitionOperation impleme
 
         if (in.getVersion().isGreaterOrEqual(Versions.V3_12)) {
             migration = in.readObject();
-            newPartitionStateVersion = in.readInt();
         } else {
             // RU_COMPAT_3_11
             partitionState = new PartitionRuntimeState();
@@ -128,7 +124,6 @@ public class MigrationCommitOperation extends AbstractPartitionOperation impleme
 
         if (out.getVersion().isGreaterOrEqual(Versions.V3_12)) {
             out.writeObject(migration);
-            out.writeInt(newPartitionStateVersion);
         } else {
             // RU_COMPAT_3_11
             partitionState.writeData(out);

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/PublishCompletedMigrationsOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/PublishCompletedMigrationsOperation.java
@@ -41,23 +41,19 @@ public class PublishCompletedMigrationsOperation extends AbstractPartitionOperat
 
     private Collection<MigrationInfo> completedMigrations;
 
-    private int newPartitionStateVersion;
-
     private transient boolean success;
 
     public PublishCompletedMigrationsOperation() {
     }
 
-    public PublishCompletedMigrationsOperation(Collection<MigrationInfo> completedMigrations,
-            int newPartitionStateVersion) {
+    public PublishCompletedMigrationsOperation(Collection<MigrationInfo> completedMigrations) {
         this.completedMigrations = completedMigrations;
-        this.newPartitionStateVersion = newPartitionStateVersion;
     }
 
     @Override
     public void run() {
         InternalPartitionServiceImpl service = getService();
-        success = service.applyCompletedMigrations(completedMigrations, newPartitionStateVersion, getCallerAddress());
+        success = service.applyCompletedMigrations(completedMigrations, getCallerAddress());
     }
 
     @Override
@@ -89,7 +85,6 @@ public class PublishCompletedMigrationsOperation extends AbstractPartitionOperat
             migrationInfo.readData(in);
             completedMigrations.add(migrationInfo);
         }
-        newPartitionStateVersion = in.readInt();
     }
 
     @Override
@@ -100,7 +95,6 @@ public class PublishCompletedMigrationsOperation extends AbstractPartitionOperat
         for (MigrationInfo migrationInfo : completedMigrations) {
             migrationInfo.writeData(out);
         }
-        out.writeInt(newPartitionStateVersion);
     }
 
     @Override

--- a/hazelcast/src/test/java/com/hazelcast/internal/partition/AbstractPartitionAssignmentsCorrectnessTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/partition/AbstractPartitionAssignmentsCorrectnessTest.java
@@ -18,8 +18,6 @@ package com.hazelcast.internal.partition;
 
 import com.hazelcast.config.Config;
 import com.hazelcast.core.HazelcastInstance;
-import com.hazelcast.instance.Node;
-import com.hazelcast.internal.cluster.impl.ClusterServiceImpl;
 import com.hazelcast.nio.Address;
 import com.hazelcast.test.AssertTask;
 import com.hazelcast.test.TestHazelcastInstanceFactory;
@@ -27,8 +25,6 @@ import org.junit.Test;
 
 import java.util.Collection;
 import java.util.Collections;
-
-import static org.junit.Assert.assertNotNull;
 
 public abstract class AbstractPartitionAssignmentsCorrectnessTest extends PartitionCorrectnessTestSupport {
 
@@ -83,30 +79,9 @@ public abstract class AbstractPartitionAssignmentsCorrectnessTest extends Partit
     static void assertPartitionAssignmentsEventually(final TestHazelcastInstanceFactory factory) {
         assertTrueEventually(new AssertTask() {
             @Override
-            public void run() throws Exception {
+            public void run() {
                 assertPartitionAssignments(factory);
             }
         });
-    }
-
-    static void assertPartitionAssignments(TestHazelcastInstanceFactory factory) {
-        Collection<HazelcastInstance> instances = factory.getAllHazelcastInstances();
-        final int replicaCount = Math.min(instances.size(), InternalPartition.MAX_REPLICA_COUNT);
-
-        for (HazelcastInstance hz : instances) {
-            Node node = getNode(hz);
-            InternalPartitionService partitionService = node.getPartitionService();
-            InternalPartition[] partitions = partitionService.getInternalPartitions();
-            ClusterServiceImpl clusterService = node.getClusterService();
-            Address thisAddress = node.getThisAddress();
-
-            for (InternalPartition partition : partitions) {
-                for (int i = 0; i < replicaCount; i++) {
-                    Address replicaAddress = partition.getReplicaAddress(i);
-                    assertNotNull("On " + thisAddress + ", Replica " + i + " is not found in " + partition, replicaAddress);
-                    assertNotNull("On " + thisAddress + ", Not member: " + replicaAddress, clusterService.getMember(replicaAddress));
-                }
-            }
-        }
     }
 }

--- a/hazelcast/src/test/java/com/hazelcast/internal/partition/MigrationInvocationsSafetyTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/partition/MigrationInvocationsSafetyTest.java
@@ -30,6 +30,7 @@ import com.hazelcast.test.ChangeLoggingRule;
 import com.hazelcast.test.HazelcastSerialClassRunner;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
+import org.hamcrest.Matchers;
 import org.junit.Before;
 import org.junit.ClassRule;
 import org.junit.Test;
@@ -89,7 +90,7 @@ public class MigrationInvocationsSafetyTest extends PartitionCorrectnessTestSupp
         dropOperationsBetween(nextMaster, singletonList(initialMaster), F_ID, singletonList(ASSIGN_PARTITIONS));
         dropOperationsBetween(slave1, singletonList(initialMaster), F_ID, singletonList(ASSIGN_PARTITIONS));
 
-        warmUpPartitions(initialMaster, slave2, slave3);
+        ensurePartitionsInitialized(initialMaster, slave2, slave3);
 
         final int initialPartitionStateVersion = getPartitionService(initialMaster).getPartitionStateVersion();
         assertEquals(initialPartitionStateVersion, getPartitionService(slave2).getPartitionStateVersion());
@@ -139,7 +140,7 @@ public class MigrationInvocationsSafetyTest extends PartitionCorrectnessTestSupp
         dropOperationsBetween(nextMaster, singletonList(initialMaster), F_ID, singletonList(ASSIGN_PARTITIONS));
         dropOperationsBetween(slave1, singletonList(initialMaster), F_ID, singletonList(ASSIGN_PARTITIONS));
 
-        warmUpPartitions(initialMaster, slave2, slave3);
+        ensurePartitionsInitialized(initialMaster, slave2, slave3);
 
         final int initialPartitionStateVersion = getPartitionService(initialMaster).getPartitionStateVersion();
         assertEquals(initialPartitionStateVersion, getPartitionService(slave2).getPartitionStateVersion());
@@ -461,5 +462,21 @@ public class MigrationInvocationsSafetyTest extends PartitionCorrectnessTestSupp
 
     private static InternalPartitionServiceImpl getPartitionServiceImpl(HazelcastInstance hz) {
         return getNode(hz).partitionService;
+    }
+
+    private static void ensurePartitionsInitialized(HazelcastInstance... instances) {
+        warmUpPartitions(instances);
+        for (HazelcastInstance instance : instances) {
+            assertPartitionStateVersionInitialized(instance);
+        }
+    }
+
+    private static void assertPartitionStateVersionInitialized(HazelcastInstance instance) {
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() {
+                assertThat(getPartitionService(instance).getPartitionStateVersion(), Matchers.greaterThan(0));
+            }
+        });
     }
 }

--- a/hazelcast/src/test/java/com/hazelcast/internal/partition/PartitionCorrectnessTestSupport.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/partition/PartitionCorrectnessTestSupport.java
@@ -19,8 +19,10 @@ package com.hazelcast.internal.partition;
 import com.hazelcast.config.Config;
 import com.hazelcast.config.ServicesConfig;
 import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.Member;
 import com.hazelcast.instance.Node;
 import com.hazelcast.instance.TestUtil;
+import com.hazelcast.internal.cluster.ClusterService;
 import com.hazelcast.internal.partition.service.TestAbstractMigrationAwareService;
 import com.hazelcast.internal.partition.service.TestIncrementOperation;
 import com.hazelcast.internal.partition.service.TestMigrationAwareService;
@@ -170,6 +172,32 @@ public abstract class PartitionCorrectnessTestSupport extends HazelcastTestSuppo
         }
     }
 
+    void assertPartitionAssignments() {
+        assertPartitionAssignments(factory);
+    }
+
+    static void assertPartitionAssignments(TestHazelcastInstanceFactory factory) {
+        Collection<HazelcastInstance> instances = factory.getAllHazelcastInstances();
+        final int replicaCount = Math.min(instances.size(), InternalPartition.MAX_REPLICA_COUNT);
+
+        for (HazelcastInstance hz : instances) {
+            Node node = getNode(hz);
+            InternalPartitionService partitionService = node.getPartitionService();
+            InternalPartition[] partitions = partitionService.getInternalPartitions();
+            ClusterService clusterService = node.getClusterService();
+            Member localMember = node.getLocalMember();
+
+            for (InternalPartition partition : partitions) {
+                for (int i = 0; i < replicaCount; i++) {
+                    PartitionReplica replica = partition.getReplica(i);
+                    assertNotNull("On " + localMember + ", Replica " + i + " is not found in " + partition, replica);
+                    assertNotNull("On " + localMember + ", Not member: " + replica,
+                            clusterService.getMember(replica.address(), replica.uuid()));
+                }
+            }
+        }
+    }
+
     void assertSizeAndDataEventually() {
         assertSizeAndDataEventually(false);
     }
@@ -178,14 +206,16 @@ public abstract class PartitionCorrectnessTestSupport extends HazelcastTestSuppo
         assertTrueEventually(new AssertSizeAndDataTask(allowDirty));
     }
 
-    void assertSizeAndData() throws InterruptedException {
+    void assertSizeAndData() {
         assertSizeAndData(false);
     }
 
-    private void assertSizeAndData(boolean allowDirty) throws InterruptedException {
+    private void assertSizeAndData(boolean allowDirty) {
         Collection<HazelcastInstance> instances = factory.getAllHazelcastInstances();
         final int actualBackupCount = Math.min(backupCount, instances.size() - 1);
         final int expectedSize = partitionCount * (actualBackupCount + 1);
+
+        assertPartitionAssignments();
 
         int total = 0;
         int[] fragmentTotals = new int[NAMESPACES.length];

